File: aliucord/src/main/java/com/aliucord/views/TextInput.java
/*
 * This file is part of Aliucord, an Android Discord client mod.
 * Copyright (c) 2021 Juby210 & Vendicated
 * Licensed under the Open Software License version 3.0
 */

package com.aliucord.views;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.*;
import android.widget.EditText;
import android.widget.LinearLayout;

import androidx.annotation.*;
import androidx.cardview.widget.CardView;
import androidx.core.content.ContextCompat;

import com.aliucord.Utils;
import com.aliucord.utils.DimenUtils;
import com.google.android.material.textfield.TextInputLayout;

public class TextInput extends CardView {
    public TextInputLayout layout;

    public TextInput(@NonNull Context context) {
        this(context, null, null, null, null);
    }

    public TextInput(@NonNull Context context, @Nullable CharSequence hint) {
        this(context, hint, null, null, null);
    }

    public TextInput(@NonNull Context context, @Nullable CharSequence hint, @Nullable String value) {
        this(context, hint, value, null, null);
    }

    public TextInput(@NonNull Context context, @Nullable CharSequence hint, @Nullable String value, @Nullable TextWatcher textChangedListener) {
        this(context, hint, value, textChangedListener, null);
    }

    public TextInput(@NonNull Context context, @Nullable CharSequence hint, @Nullable String value, @Nullable View.OnClickListener endIconOnClick) {
        this(context, hint, value, null, endIconOnClick);
    }

    public TextInput(@NonNull Context context, @Nullable CharSequence hint, @Nullable String value, @Nullable TextWatcher textChangedListener, @Nullable View.OnClickListener endIconOnClick) {
        super(context);
        LinearLayout root = new LinearLayout(context);
        LayoutInflater.from(context).inflate(Utils.getResId("widget_change_guild_identity", "layout"), root);
        layout = (TextInputLayout) root.findViewById(Utils.getResId("set_nickname_text", "id"));
        ((CardView) layout.getParent()).removeView(layout);
        addView(layout);
        setCardBackgroundColor(Color.TRANSPARENT);
        getRoot().setHint(hint == null ? "Enter Text" : hint);
        if(value != null && !value.isEmpty()) getEditText().setText(value);
        //if(placeholder != null && !placeholder.isEmpty()) getEditText().setPlaceholder(placeholder);
        setRadius(DimenUtils.getDefaultCardRadius());
        getRoot().setEndIconVisible(false);
        getEditText().addTextChangedListener(textChangedListener == null ? new TextWatcher() {
            public void afterTextChanged(Editable s) {
                getRoot().setEndIconVisible(!s.toString().isEmpty());
            }

            public void beforeTextChanged(CharSequence s, int start, int count, int after) { }
            public void onTextChanged(CharSequence s, int start, int before, int count) { }
        } : textChangedListener);
        getRoot().setEndIconOnClickListener(endIconOnClick == null ? v -> getEditText().setText("") : endIconOnClick);
    }

    /**
     * Returns the root layout
     * @return TextInputLayout
     */
    @NonNull
    public TextInputLayout getRoot() {
        return layout;
    }

    /**
     * Returns the main edit text
     * @return EditText
     */
    @NonNull
    public EditText getEditText() {
        return (EditText) ((ViewGroup) getRoot().getChildAt(0)).getChildAt(0);
    }

    /**
     * Sets the hint message
     * @param hint The hint
     * @return self
     * @noinspection UnusedReturnValue
     */
    public TextInput setHint(@NonNull CharSequence hint) {
        getRoot().setHint(hint);
        return this;
    }

    /**
     * Sets the hint message
     * @param hint The hint res id
     * @return self
     */
    public TextInput setHint(@StringRes int hint) {
        getRoot().setHint(hint);
        return this;
    }

    /**
     * Sets the end icon to the specified drawable and sets it tint to the users's chosen theme
     * @param icon End icon drawable
     * @return self
     */
    public TextInput setThemedEndIcon(@NonNull Drawable icon) {
        getRoot().setEndIconDrawable(Utils.tintToTheme(icon.mutate()));
        return this;
    }

    /**
     * Sets the end icon to the specified drawable and sets it tint to the users's chosen theme
     * @param icon End icon drawable res id
     * @return self
     */
    public TextInput setThemedEndIcon(@DrawableRes int icon) {
        getRoot().setEndIconDrawable(Utils.tintToTheme(ContextCompat.getDrawable(getRoot().getContext(), icon)));
        return this;
    }
}

File: Aliucord/src/main/java/com/aliucord/views/TextInput.kt
/*
 * This file is part of Aliucord, an Android Discord client mod.
 * Copyright (c) 2021 Juby210 & Vendicated
 * Licensed under the Open Software License version 3.0
 */
package com.aliucord.views

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.text.Editable
import android.text.TextWatcher
import android.view.*
import android.view.View.OnClickListener
import android.widget.EditText
import android.widget.LinearLayout
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.cardview.widget.CardView
import androidx.core.content.ContextCompat
import com.aliucord.Utils.getResId
import com.aliucord.Utils.tintToTheme
import com.aliucord.utils.DimenUtils.defaultCardRadius
import com.google.android.material.textfield.TextInputLayout


class TextInput @JvmOverloads constructor(
    context: Context,
    hint: CharSequence? = null,
    value: String? = null,
    textChangedListener: TextWatcher? = null,
    endIconOnClick: OnClickListener? = null
) : CardView(context) {
    var layout: TextInputLayout

    constructor(
        context: Context,
        hint: CharSequence?,
        value: String?,
        endIconOnClick: OnClickListener?
    ) : this(context, hint, value, null, endIconOnClick)

    /**
     * Returns the main edit text
     * @return EditText
     */
    val editText: EditText
        get() = (layout.getChildAt(0) as ViewGroup).getChildAt(0) as EditText

    /**
     * Returns the root layout
     * @return TextInputLayout
     */
    fun getRoot(): TextInputLayout {
        return layout
    }

    /**
     * Sets the hint message
     * @param hint The hint
     * @return self
     */
    fun setHint(hint: CharSequence): TextInput {
        layout.hint = hint
        return this
    }

    /**
     * Sets the hint message
     * @param hint The hint res id
     * @return self
     */
    fun setHint(@StringRes hint: Int): TextInput {
        layout.setHint(hint)
        return this
    }

    /**
     * Sets the end icon to the specified drawable and sets it tint to the users's chosen theme
     * @param icon End icon drawable
     * @return self
     */
    fun setThemedEndIcon(icon: Drawable): TextInput {
        layout.endIconDrawable = tintToTheme(icon.mutate())
        return this
    }

    /**
     * Sets the end icon to the specified drawable and sets it tint to the users's chosen theme
     * @param icon End icon drawable res id
     * @return self
     */
    fun setThemedEndIcon(@DrawableRes icon: Int): TextInput {
        layout.endIconDrawable =
            tintToTheme(ContextCompat.getDrawable(layout.context, icon))
        return this
    }

    init {
        val root = LinearLayout(context)
        LayoutInflater.from(context)
            .inflate(getResId("widget_change_guild_identity", "layout"), root)
        layout = root.findViewById<View>(getResId("set_nickname_text", "id")) as TextInputLayout
        (layout.parent as CardView).removeView(layout)
        addView(layout)
        setCardBackgroundColor(Color.TRANSPARENT)
        getRoot().hint = hint ?: "Enter Text"
        if (value != null && value.isNotEmpty()) editText.setText(value)
        //if(placeholder != null && !placeholder.isEmpty()) getEditText().setPlaceholder(placeholder);
        radius = defaultCardRadius.toFloat()
        getRoot().isEndIconVisible = false

        editText.addTextChangedListener(textChangedListener
            ?: object : TextWatcher {
                override fun afterTextChanged(s: Editable) {
                    getRoot().isEndIconVisible = s.toString().isNotEmpty()
                }

                override fun beforeTextChanged(
                    s: CharSequence,
                    start: Int,
                    count: Int,
                    after: Int
                ) {
                }

                override fun onTextChanged(
                    s: CharSequence,
                    start: Int,
                    before: Int,
                    count: Int
                ) {
                }
            })
        getRoot().setEndIconOnClickListener(
            endIconOnClick ?: OnClickListener { editText.setText("") })
    }
}
